honeybadgermpc/protocols/AVID.py

```diff
@@ -0,0 +1,140 @@
+            (i, (tag, ...)) = receive()
+        where ``tag`` is one of ``{"VAL", "ECHO", "READY"}``
+        :param send: sends (without blocking) a message to a designed
+        recipient ``send(i, (tag, ...))``
+        """
+        assert n >= 3 * f + 1
+        assert f >= 0
+        assert 0 <= leader < n
+
+        self.__data = [None] * n
+        self.receive = receive
+        self.send = send
+        self.n = n
+        self.f = f
+        self.leader = leader
+
+    def retrieve(self, i, k=None):
+        """ Retrieve individual item
+        :param int i:  the item for party i
+        :param int k: TODO the k'th secret
+        """
+        # we are able to retrieve things from AVID indivdually
+        assert 0 <= i < self.n
+        return self.__data[i]
+
+    async def disperse(self, sid, pid, input_list):
+        """ Main information dispersal handling
+        :param int sid: e.x. tag to be used
+        :param int pid: current member id
+        :param int input_list: the list of message
+            for each party from 1 to n from the dealer
+        """
+        # setup some parameters
+        k = self.n - 2 * self.f  # Wait to reconstruct.
+        echo_threshold = math.ceil((self.n + self.f + 1) / 2)  # number of ECHOs needed
+        ready_threshold = self.f + 1  # number of READYs needed for READY
+        output_threshold = 2 * self.f + 1  # number of READYs needed for storing
+
+        def broadcast(o):
+            for i in range(self.n):
+                self.send(i, o)
+
+        # leader handling
+        if pid == self.leader:
+            # The leader erasure encodes the input, sending one strip to each participant
+            m_list = input_list  # block until an input list is received
+            assert len(m_list) == self.n
+
+            # construct the stripe list which contains the stripes for party 1 to party n
+            stripes_list = [None] * self.n
+            for i in range(len(m_list)):
+                m = m_list[i]
+                stripes_list[i] = encode(k, self.n, m)
+
+            for i in range(self.n):
+                # send each person the column of stripes
+                stripes_for_each = [stripes[i] for stripes in stripes_list]
+                self.send(i, (sid, 'VAL', stripes_for_each))
+
+        result = [None] * self.n  # the result current party got from all parties
+        echo_set = set()
+        ready_set = set()
+        ready_sent = False
+        from_leader = None
+
+        def decode_output():
+            m = decode(k, self.n, result)
+            return m
+
+        while True:  # main receive loop
+            sender, msg = await self.receive()
+            if msg[1] == 'VAL' and from_leader is None:
+                # Validation
+                # TODO merkle tree validation
+                (_, _, stripes_for_each) = msg
+                if sender != self.leader:
+                    logger.info(f"[{pid}] VAL message from other than leader: {sender}")
```

Comment:
log these as warnings and not info messages

Reply:
Same comment below

Reply:
resolved. Same case in reliable broadcast.